src/Microsoft.AspNetCore.Routing/Matchers/TreeMatcher.cs

```diff
@@ -0,0 +1,348 @@
+            feature.Values = values;
+
+            var cache = _cache.Value;
+            for (var i = 0; i < cache.Length; i++)
+            {
+                var tree = cache[i];
+                var tokenizer = new PathTokenizer(httpContext.Request.Path);
+
+                var treenumerator = new Treenumerator(tree.Root, tokenizer);
+
+                while (treenumerator.MoveNext())
+                {
+                    var node = treenumerator.Current;
+                    foreach (var item in node.Matches)
+                    {
+                        var entry = item.Entry;
+                        var matcher = item.TemplateMatcher;
+
+                        values.Clear();
+                        if (!matcher.TryMatch(httpContext.Request.Path, values))
+                        {
+                            continue;
+                        }
+
+                        Log.MatchedTemplate(_logger, httpContext, entry.RouteTemplate);
+
+                        if (!MatchConstraints(httpContext, values, entry.Constraints))
+                        {
+                            continue;
+                        }
+
+                        await SelectEndpointAsync(httpContext, feature, (MatcherEndpoint[])entry.Tag);
+
+                        if (feature.Endpoint != null)
+                        {
+                            if (feature.Endpoint is MatcherEndpoint endpoint)
+                            {
+                                foreach (var kvp in endpoint.Values)
+                                {
+                                    if (!feature.Values.ContainsKey(kvp.Key))
+                                    {
+                                        feature.Values[kvp.Key] = kvp.Value;
+                                    }
+                                }
+                            }
+
+                            return;
+                        }
+                    }
+                }
+            }
+        }
+
+        private bool MatchConstraints(
+            HttpContext httpContext,
+            RouteValueDictionary values,
+            IDictionary<string, IRouteConstraint> constraints)
+        {
+            if (constraints != null)
+            {
+                foreach (var kvp in constraints)
+                {
+                    var constraint = kvp.Value;
+                    if (!constraint.Match(httpContext, null, kvp.Key, values, RouteDirection.IncomingRequest))
+                    {
+                        values.TryGetValue(kvp.Key, out var value);
+
+                        Log.ConstraintFailed(_logger, value, kvp.Key, kvp.Value);
+                        return false;
+                    }
+                }
+            }
+
+            return true;
+        }
+
+        private Task SelectEndpointAsync(HttpContext httpContext, IEndpointFeature feature, IReadOnlyList<MatcherEndpoint> endpoints)
+        {
+            switch (endpoints.Count)
```

Comment:
Note that this code doesn't do anything significant now. This will eventually incorporate something like `IActionConstraint`.